skimage/filters/thresholding.py

```diff
@@ -944,20 +944,27 @@ def _mean_std(image, w):
     pad_width = tuple((k // 2 + 1, k // 2) for k in w)
     padded = np.pad(image.astype('float'), pad_width,
                     mode='reflect')
-    padded_sq = padded * padded
 
     integral = integral_image(padded)
-    integral_sq = integral_image(padded_sq)
-
-    kern = np.zeros(tuple(k + 1 for k in w))
-    for indices in itertools.product(*([[0, -1]] * image.ndim)):
-        kern[indices] = (-1) ** (image.ndim % 2 != np.sum(indices) % 2)
+    padded *= padded
+    integral_sq = integral_image(padded)
+
+    # Store the kernel as a list 2-tuples where:
+    #     - The first element is an index into the kernel (of shape w).
+    #     - The second element is the value at that index.
+    kernel_indices_and_values = []
+    for indices in itertools.product(*tuple([(0, _w) for _w in w])):
+        kernel_indices_and_values.append(
+            (indices, (-1) ** (image.ndim % 2 != np.sum(indices) % 2))
+        )
 
     total_window_size = np.prod(w)
-    sum_full = correlate_sparse(integral, kern, mode='valid')
-    m = sum_full / total_window_size
-    sum_sq_full = correlate_sparse(integral_sq, kern, mode='valid')
-    g2 = sum_sq_full / total_window_size
+    kernel_shape = tuple(_w + 1 for _w in w)
+    m = _correlate_sparse(integral, kernel_shape, kernel_indices_and_values)
```

Comment:
As a minor point, from an API niceness standpoint I'd rather `kernel_indices` and `kernel_values` be two separate arrays that are zipped inside the function. Not critical though.